src/Common/tests/Collections/ICollectionTest.cs

```diff
@@ -0,0 +1,288 @@
+            }
+        }
+
+        protected ICollectionTest(bool isSynchronized)
+        {
+            ExpectedIsSynchronized = isSynchronized;
+            ValidArrayTypes = new[] { typeof(object) };
+            InvalidArrayTypes = new[]
+            {
+                typeof (MyInvalidReferenceType),
+                typeof (MyInvalidValueType)
+            };
+        }
+
+        [Fact]
+        public void Count()
+        {
+            var items = GenerateItems(16);
+            var collection = GetCollection(items);
+            Assert.Equal(items.Length, collection.Count);
+        }
+
+        [Fact]
+        public void IsSynchronized()
+        {
+            var collection = GetCollection(GenerateItems(16));
+            Assert.Equal(ExpectedIsSynchronized, collection.IsSynchronized);
+        }
+
+        [Fact]
+        public void SyncRootNonNull()
+        {
+            var collection = GetCollection(GenerateItems(16));
+            Assert.NotNull(collection.SyncRoot);
+        }
+
+        [Fact]
+        public void SyncRootConsistent()
+        {
+            var collection = GetCollection(GenerateItems(16));
+            var syncRoot1 = collection.SyncRoot;
+            var syncRoot2 = collection.SyncRoot;
+            Assert.Equal(syncRoot1, syncRoot2);
+        }
+
+        [Fact]
+        public void SyncRootCanBeLocked()
+        {
+            var collection = GetCollection(GenerateItems(16));
+            lock (collection.SyncRoot)
+            {
+            }
+        }
+
+        [Fact]
+        public void SyncRootUnique()
+        {
+            var collection1 = GetCollection(GenerateItems(16));
+            var collection2 = GetCollection(GenerateItems(16));
+            Assert.NotEqual(collection1.SyncRoot, collection2.SyncRoot);
+        }
+
+        [Fact]
+        public void CopyToNull()
+        {
+            var collection = GetCollection(GenerateItems(16));
+            Assert.Throws<ArgumentNullException>(
+                () => collection.CopyTo(null, 0));
+        }
+
+        [Theory]
+        [InlineData(int.MinValue)]
+        [InlineData(-1)]
+        [InlineData(int.MaxValue)]
+        public void CopyToBadIndex(int index)
+        {
+            var items = GenerateItems(16);
+            var collection = GetCollection(items);
+            var items2 = (object[]) items.Clone();
+
+            Assert.ThrowsAny<ArgumentException>(
+                () => collection.CopyTo(items2, index));
+
+            CollectionAssert.Equal(items, items2);
+        }
+
+        [Fact]
+        public void CopyToArrayWithNonZeroBounds()
+        {
+            var items = GenerateItems(16);
+            var collection = GetCollection(items);
+            if (CopyToOnlySupportsZeroLowerBounds)
+            {
+                var itemArray = Array.CreateInstance(
+                    typeof (object),
+                    new[] {collection.Count + 8},
+                    new[] {-4});
```

Comment:
agreed there isn't any good reason to use Array.CreateInstance with reflection for these cases.